### PR TITLE
Move preallocate functionality to native access

### PR DESCRIPTION
**Pull request by rjernst**
This commit moves the file preallocation functionality into NativeAccess. The code is basically the same. One small tweak is that instead of breaking Java access boundaries in order to get an open file handle, the new code uses posix open directly.

relates #104876

### CODE REVIEW

File: libs/native/src/main/java/org/elasticsearch/nativeaccess/PosixNativeAccess.java [hunk 1]
```diff
@@ -121,6 +125,49 @@ public void tryLockMemory() {
...
+        assert Files.isRegularFile(path) : path;
+        var stats = libc.newStat64(constants.statStructSize(), constants.statStructSizeOffset(), constants.statStructBlocksOffset());
+
+        int fd = libc.open(path.toAbsolutePath().toString(), O_WRONLY, constants.O_CREAT());
+        if (fd == -1) {
+            logger.warn("Could not open file [" + path + "] to get allocated size: " + libc.strerror(libc.errno()));
+            return OptionalLong.empty();
+        }
+
+        if (libc.fstat64(fd, stats) != 0) {
```
Comment (ldematte) [inline libs/native/src/main/java/org/elasticsearch/nativeaccess/PosixNativeAccess.java:139]: This is failing in the tests; in the original code, xstat was used.
I don't think this is relevant - maybe xstat can get a file size for a file that was just pre-allocated, and fstat cannot, but I seriously doubt it.
The errno is `Invalid Argument`. I'd double check the offsets for the struct here.

Comment (rjernst) [reply libs/native/src/main/java/org/elasticsearch/nativeaccess/PosixNativeAccess.java:140]: This was a bug in how the file was opened. I copy/pasted the open code and used O_WRONLY, when it should have been O_RDONLY.

Comment (rjernst) [reply libs/native/src/main/java/org/elasticsearch/nativeaccess/PosixNativeAccess.java:140]: Hrm, or at least that's what I thought. I got bit by an assume that was making this test not run locally. I'll keep digging.

Comment (ldematte) [reply libs/native/src/main/java/org/elasticsearch/nativeaccess/PosixNativeAccess.java:140]: I see you switched to Memory; good call. It seems that even if you use a JNA `Structure` byte[] are not inlined in the struct, so the padding was not right. Better use a block of memory and access the correct offset.

File: x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/cache/common/CacheFileTests.java [hunk 2]
```diff
@@ -393,15 +393,7 @@ public void testFSyncFailure() throws Exception {
...
 
-    private static void assumeLinux64bitsOrWindows() {
-        assumeTrue(
-            "This test uses native methods implemented only for Windows & Linux 64bits",
-            Constants.WINDOWS || Constants.LINUX && Constants.JRE_IS_64BIT
-        );
-    }
-
     public void testCacheFileCreatedAsSparseFile() throws Exception {
-        assumeLinux64bitsOrWindows();
```
Comment (ldematte) [inline x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/cache/common/CacheFileTests.java:404]: I see you are removing this, even if the pre-allocate for Windows is a no-op; I think you are right, a quick glance at the test seems to indicate the assertions would be working on Windows too. Just to be safe, I suggest to make a final test run with `test-windows`.

**Review by ldematte — APPROVED** [review]
I see a failing test in serverless; it could be unrelated, I gave a quick look but it's not obvious. 
Otherwise, LGTM

**Comment by rjernst** [comment]
> I see a failing test in serverless; it could be unrelated

I think it is unrelated. I added serverless testing because there is an existing issue (on main) around preallocate failing in jna init. I wanted to see if this PR still had that problem, which it appears not to have.

**Review by breskeby — COMMENTED** [review]
fyi: https://github.com/elastic/infra/pull/41298

**Pull request merged**